presto-main/src/main/java/com/facebook/presto/spiller/LocalTempStorage.java

```diff
@@ -72,11 +72,10 @@ public LocalTempStorage(List<Path> spillPaths, double maxUsedSpaceThreshold)
     {
         this.spillPaths = ImmutableList.copyOf(requireNonNull(spillPaths, "spillPaths is null"));
         this.maxUsedSpaceThreshold = maxUsedSpaceThreshold;
+        initialize();
```

Comment:
Is there any benefit of late initialization like avoid creation of directories for spill paths in constructor?

Reply:
Just for the context.

When testing broadcast join implementation @pgupta2 discovered that the temporary files are randomly disappearing. Long story short - if the storage service is used by multiple components it is hard to ensure it is initialized only once, and that it is always initialized before usage. 

The lazy initialization can still be implemented by wrapping the storage into a "Lazy" wrapper, a common pattern that we use in other places across the code base.

Reply:
I agree with the idea that initialization should happen internally as per the storage requirements. However, there could be scenarios where a specific use-case would want to perform some setup before using the storage. The initialize() API provided that flexibility in some sense.

Does it make sense to add a new API like prepare() or setup() that can be used to prepare the storage for a given use case. Ex: For broadcast join, we would prefer to setup a temp directory and use that location for writing all spill files. Or probably use separate directory for each query. With current setup, all spill files are written to same spill directories and it makes it tricky to debug/investigate.

Reply:
> However, there could be scenarios where a specific use-case would want to perform some setup before using the storage. The initialize() API provided that flexibility in some sense.

Yeah, that's why I thought originally that the `initialize()` method in the interface is a good idea.

Yesterday I spend quite some time analyzing the design, and it feels like this flexibility can be achieved by registering multiple storage services with a different configuration.

Maybe let me quickly recap the current design:

1. Plugins can provide storage service implementation by providing a `TempStorageFactory`. All available `TempStorageFactory` are registered in the `TempStorageManager`, that is a singleton: https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/server/PluginManager.java#L258
2. 1TempStorageFactory` can create a temporary storage based on a configuration. A single `TempStorageFactory` (for example `HdfsTempStorageFactory`, or `LocalTempStorageFactory`) can create multiple storage instances. For example the `LocalTempStorageFactory` can create one instance that stores temporary files under the `/tmp/folder_1` and a second instance that stores temporary files under the `/tmp/folder_2`.
3. `TempStorage` instances are initialized based on the configuration in the `TempStorageManager`: https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/storage/TempStorageManager.java#L117. In the configuration we specify what storage implementation to use along with the storage configuration (e.g.: where to store the temporary files). The initialized `TempStorage` service is then registered under a specified name (e.g.: `local_storage_folder_1`, `local_storage_folder_2`).
4. The `TempStorageManager` can return an instance by its name: https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/storage/TempStorageManager.java#L109. It allows different Presto components to be configured to use different storage instances. For example spilling can be configured to use the `local_storage_folder_1` and broadcast can be configured to use the `local_storage_folder_2`.
5. Both instances, the `local_storage_folder_1` and the `local_storage_folder_2` are singletons, that could potentially be used by multiple independent components (for example spilling and broadcast), thus it is unclear what components should be responsible for the initialization (especially if different components want to configure the same storage service instance differently)

Given this design

> For broadcast join, we would prefer to setup a temp directory and use that location for writing all spill files.

can be implemented by having 2 separate storage instances, one for broadcast, one for spilling

> Or probably use separate directory for each query

And this might be problematic, since the storage service is currently designed to be a singleton.

Reply:
> Or probably use separate directory for each query

And this might be problematic, since the storage service is currently designed to be a singleton.

Actually an afterthought (based on the discussion offline)

Separate directory per query can in fact be implemented currently, as the query id is available in the `TempDataOperationContext`.